warehouse/oidc/views.py

```diff
@@ -0,0 +1,88 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+
+from pyramid.view import view_config
+
+from warehouse.macaroons.interfaces import IMacaroonService
+from warehouse.oidc.interfaces import IOIDCProviderService
+
+
+@view_config(
+    route_name="oidc.mint_token",
+    require_methods=["POST"],
+    renderer="json",
+    require_csrf=False,
+    has_translations=False,
+)
+def mint_token_from_oidc(request):
+    def _invalid(errors):
+        request.response.status = 422
+        return {"message": "Token request failed", "errors": errors}
+
+    try:
+        body = request.json_body
+    except ValueError:
+        return _invalid(
+            errors=[{"code": "invalid-json", "description": "missing JSON body"}]
+        )
+
+    unverified_jwt = body.get("token")
+    if not unverified_jwt:
+        return _invalid(
+            errors=[{"code": "invalid-token", "description": "missing or empty token"}]
+        )
+
+    # For the time being, GitHub is our only OIDC provider.
+    # In the future, this should locate the correct service based on an
+    # identifier in the request body.
+    oidc_service = request.find_service(IOIDCProviderService, name="github")
+    provider = oidc_service.find_provider(unverified_jwt)
+    if not provider:
+        return _invalid(
+            errors=[
+                {"code": "invalid-token", "description": "malformed or invalid token"}
+            ]
+        )
+
+    # At this point, we've verified that the given JWT is valid for the given
+    # project. All we need to do is mint a new token.
+    macaroon_service = request.find_service(IMacaroonService, context=None)
+    now = time.time()
+    expires = int(now) + 900
+    caveats = [
+        {"permissions": "oidc", "version": 1},
+        # TODO: Does this caveat make sense anymore?
+        # Conceptually, an OIDC provider authorizes either for every project registered
+        # against it at the time of token creation, *or* for every project that happens
+        # to be registered against it at the time of use. This caveat constrains the
+        # token to the former, but maybe the latter is fine/expected?
+        {"project_ids": [str(p.id) for p in provider.projects]},
```

Comment:
Flagging for review (cc @dstufft in particular).

Reply:
I'm going to remove this TODO -- it's stale with the caveat refactor, and I'm convinced that it makes sense for an API token to be limited to only the projects that existed at the time it was created.